lib/openapi_contracts/validators/request.rb

```diff
@@ -0,0 +1,42 @@
+module OpenapiContracts::Validators
+  class Request < Base
+    private
+
+    def validate
+      validate_schema if match_request_body?
+    end
+
+    def validate_schema
+      if !request_body
+        @errors << 'No request body found'
+      elsif !request_body.supports_content_type?(request_content_type)
+        @errors << "Undocumented request with content-type #{request_content_type.inspect}"
+      else
+        schema = request_body.schema_for(request_content_type)
+        # Trick JSONSchemer into validating only against the request body schema
+        schemer = JSONSchemer.schema(schema.raw.merge('$ref' => schema.fragment, '$schema' => 'http://json-schema.org/draft-04/schema#'))
+        schemer.validate(JSON(request_payload)).each do |err|
+          @errors << error_to_message(err)
+        end
+      end
+    end
+
+    def request_payload
+      response.request.env['PARAMS'] || response.request.body.read
```

Comment:
Btw where did you encounter this `PARAMS` env variable? Was the default `response.request.body.read` not working in that case?
Also in case of `response.request.body.read` it might be safer to do a `rewind` first.

Reply:
The default `response.request.body.read` was not working in the specs of the gem.

I was having problems with the `Rack::Request` we are using to mock requests. I wasn't able to find a way to read the body directly from the Rack request, but only by reading the key `PARAMS` from its `env` variable.

So, basically, even though the request validation works fine without `response.request.env['PARAMS']` in my application, I have added it for the cases where Rack is being used. 

I'd love to hear your opinion/solution to this problem.

Reply:
Ah i see you injected `PARAMS` in the tests.

I added a fix on main branch that improves the usage of Rack Request & Response. Can you merge it into your branch & fix conflicts.

Then you can mock the request body by using `TestRequest.build(path, method: method, input: request_body)`, see the updated `setup_context.rb`.